### PR TITLE
add DataTransfer

### PR DESCRIPTION
**Pull request by glenn-lish**

### CODE REVIEW

File: DataTransfer/src/main/java/com/kingland/eip/UI.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+
+public class UI {
+
+    static ExecutorService es = Executors.newFixedThreadPool(2);
+    static Queue<List<User>> cache = new LinkedBlockingQueue<>();
+
+    public static void create(){
```
Comment (lewisliu725) [inline DataTransfer/src/main/java/com/kingland/eip/UI.java:25]: PR2: The method is too large, a method should contains no more code than one screen can afford.

File: DataTransfer/src/main/java/com/kingland/eip/UI.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                                    clcs_lock.wait();
+                                } catch (InterruptedException e) {
+                                    e.printStackTrace();
+                                }
+                            }
+                            clcs_lock.notifyAll();
+                            UI.displayCurrentSize(cache.size());
+                        }
+                    }
+                });
```
Comment (lewisliu725) [inline DataTransfer/src/main/java/com/kingland/eip/UI.java:71]: PR1: The synchronized code block disabled the advantage of the multiple thread, because of the synchronized code code, the Loader and Reader wouldn't be run in parallel. So there is no differenece than single thread.

File: DataTransfer/src/main/java/com/kingland/eip/UI.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                                    flcs_lock.wait();
+                                } catch (InterruptedException e) {
+                                    e.printStackTrace();
+                                }
+                            }
+                            flcs_lock.notifyAll();
+                            UI.displayCurrentSize(cache.size());
+                        }
+                    }
+                });
```
Comment (lewisliu725) [inline DataTransfer/src/main/java/com/kingland/eip/UI.java:100]: PR2: Actually, this code is duplicate with the first case, from an abstraction perspective, they can be the same.

Same comment for other cases.

File: DataTransfer/src/main/java/com/kingland/eip/loader/ConsoleDataLoader.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * Console data loader
+ */
+public class ConsoleDataLoader implements DataLoader {
+
+    public ConsoleDataLoader() {}
+
+    @Override
+    public List<User> load() {
+        List<User> users = new ArrayList<>();
+        for(int i=1;i<=10;i++)
```
Comment (lewisliu725) [inline DataTransfer/src/main/java/com/kingland/eip/loader/ConsoleDataLoader.java:21]: It would be better to have the consumer to control the batch size since it's limited by the buffer size.

File: DataTransfer/src/main/java/com/kingland/eip/loader/DataLoader.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+ * Copyright 2020 Kingland Systems Corporation. All Rights Reserved.
+ */
+package com.kingland.eip.loader;
+
+/**
+ * Data Loader
+ */
+public interface DataLoader {
+
+    Object load();
```
Comment (lewisliu725) [inline DataTransfer/src/main/java/com/kingland/eip/loader/DataLoader.java:11]: PR2: The `load` method is expected to load a list of items, so it should return a list or collection of items instead a single `Object`.

And use `Object` is not recommaned, it would requite the consumer to use the type cast, which may cause lots of unepxected issues. So please use generic type instead.

File: DataTransfer/src/main/java/com/kingland/eip/sender/DataSender.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import com.kingland.eip.User;
+
+import java.util.List;
+
+/**
+ * Data Sender
+ */
+public interface DataSender {
+
+    boolean send(Object loadData);
```
Comment (lewisliu725) [inline DataTransfer/src/main/java/com/kingland/eip/sender/DataSender.java:15]: Same comment as `DataLoader`.

File: DataTransfer/src/test/java/com/kingland/eip/DataCacheTest.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.util.Queue;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class DataCacheTest {
+
+    @Test
+    void shouldEqualEvenIfMultipleDeclared() {
+        // Given
```
Comment (lewisliu725) [inline DataTransfer/src/test/java/com/kingland/eip/DataCacheTest.java:14]: It's good to add a brief introduction about what's `given`, the same comment for `when` and `then`.

File: DataTransfer/src/test/java/com/kingland/eip/DataCacheTest.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package com.kingland.eip;
+
+import com.kingland.eip.sender.ConsoleDataSender;
+import org.junit.jupiter.api.Test;
+
+import java.util.Queue;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class DataCacheTest {
```
Comment (lewisliu725) [inline DataTransfer/src/test/java/com/kingland/eip/DataCacheTest.java:10]: PR1: I don't think the test coverage is enough, so you may need to create more unit tests.

**Review by lewisliu725 — CHANGES_REQUESTED** [review]
